Improve memory read/write

qwe661234 opened the pull request:
After analyzing the data collected during the execution of the Dhrystone benchmark, we found that the primary performance bottleneck lies in memory read/write operations. To address this issue, we modified the implementation of memory read/write by eliminating unnecessary checks and replacing slow operations with more efficient ones. Additionally, we integrated a simple memory pool and resolved the issue of disabled unaligned memory access, the user can add argument `--misalign` to enable misaligned memory access.

After analyzing the performance results from running Dhrystone, we observed a significant improvement following our enhancements to the memory I/O. 

| Test | 66200d0 | improvement | Speedup |
|--------| -------- | -------- |-------- |
| Dhrystone  | 815 DMIPS    |  1245 DMIPS | +52.7% |

See: #122

jserv commented on src/riscv_private.h:
```diff
@@ -119,6 +119,8 @@ struct riscv_internal {
     bool compressed;       /**< current instruction is compressed or not */
     block_map_t block_map; /**< basic block map */
+
+    bool is_misaliged_enabled; /**< enable memory misaligned access */
```
Rename to `allow_ misalign`.

jserv reviewed (changes requested):
Since the misaligned memory access is optional, we don't have to check it for the internals of emulator. Instead, we can register the callback functions while initializing the emulator.

See [async.h](https://github.com/embedded2016/server-framework/blob/master/async.h) and [async.c](https://github.com/embedded2016/server-framework/blob/master/async.c) for details.
```c
/* API gateway */
struct __ASYNC_API__ Async = {
    .create = async_create,
    .signal = async_signal,
    .wait = async_wait,
    .finish = async_finish,
    .run = async_run,
};
```

Once you prepare the functions specialized for both misaligned memory and strictly-aligned memory access, the proper I/O implementations can be hooked up. Consequently, you don't have to check `allow_ misalign` all the time.

jserv commented on src/io.c:
```diff
@@ -125,14 +109,46 @@ void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
+    const uint32_t addr_lo = addr & mask_lo, addr_hi = addr >> 16;
+    chunk_t *c = m->chunks[addr_hi];
+    if (unlikely(!c)) {
+        c = calloc(1, sizeof(chunk_t));
```
Can you eliminate the use of `calloc` here?

qwe661234 replied:
If we want to eliminate it, we should know the total memory size and allocate it first.

jserv replied:
Let's introduce memory pool within the pull request.

jserv replied:
You don't have to integrate TLSF (or its variants such as xvmalloc) within this commit. Instead, an arbitrary memory pool works for this case.

qwe661234 replied:
I roughly integrate the [TLSF 2.4.6](http://www.gii.upv.es/tlsf/main/repo) to file `io.c`.

jserv commented on src/main.c:
```diff
@@ -202,7 +211,7 @@ int main(int argc, char **args)
 
     /* create the RISC-V runtime */
-    riscv_t *rv = rv_create(&io, state);
+    riscv_t *rv = rv_create(&io, state, opt_misaligned);
```
Don't change the prototype of `rv_create()`. Instead, consolidate `state` or `io`.

jserv commented:
> | Test	| 66200d0	| Proposed |	Speedup |
> |-------|--------------|-----------|-------------|
> | Dhrystone	| 815 DMIPS	| 1200 DMIPS	| +47.2% |

What else do you consider it would be beneficial to employ in place of Spike for faster interpreting?

qwe661234 commented:
> Since the misaligned memory access is optional, we don't have to check it for the internals of emulator. Instead, we can register the callback functions while initializing the emulator.
> 
> See [async.h](https://github.com/embedded2016/server-framework/blob/master/async.h) and [async.c](https://github.com/embedded2016/server-framework/blob/master/async.c) for details.
> 
> ```c
> /* API gateway */
> struct __ASYNC_API__ Async = {
>     .create = async_create,
>     .signal = async_signal,
>     .wait = async_wait,
>     .finish = async_finish,
>     .run = async_run,
> };
> ```
> 
> Once you prepare the functions specialized for both misaligned memory and strictly-aligned memory access, the proper I/O implementations can be hooked up. Consequently, you don't have to check `allow_ misalign` all the time.

But the memory I/O implemention for for both misaligned memory and strictly-aligned memory access is the same, the only difference is that we need to handle exception if misaligned memory access is disabled.

jserv commented:
> the memory I/O implemention for for both misaligned memory and strictly-aligned memory access is the same, the only difference is that we need to handle exception if misaligned memory access is disabled.

Then, you can hook exception handlers which are aware of misaligned memory access by means of runtime dispatch.

qwe661234 commented:
> > the memory I/O implemention for for both misaligned memory and strictly-aligned memory access is the same, the only difference is that we need to handle exception if misaligned memory access is disabled.
> 
> Then, you can hook exception handlers which are aware of misaligned memory access by means of runtime dispatch.

Does this means send a callback function to memory read/ write? Because we don't need load/store handler if the memory misaligned access is enabled.

jserv commented:
> > Then, you can hook exception handlers which are aware of misaligned memory access by means of runtime dispatch.
> Does this means send a callback function to memory read/ write?

Sort of. In particular, we can shrink the code below:
```c
/* SW: Store Word */
RVOP(sw, {
    const uint32_t addr = rv->X[ir->rs1] + ir->imm;
    if (!rv->is_misaliged_enabled && unlikely(addr & 3)) {
        rv->compressed = false;
        rv_except_store_misaligned(rv, addr);
        return false;
    ...
```
by replacing with something like the following:
```c
/* SW: Store Word */
RVOP(sw, {
    const uint32_t addr = rv->X[ir->rs1] + ir->imm;
    RV_EXC_MISAGLIN_HANDLER(); /* some macro */
```

Meanwhile, we can hide the implementation details about `allow_ misalign` checks.

jserv commented on src/riscv.h:
```diff
@@ -107,6 +107,9 @@ typedef struct {
     riscv_on_ecall on_ecall;
     riscv_on_ebreak on_ebreak;
+
+    /* enable memory misaligned access */
```
Should be "enable misaligned memory access"

jserv commented on src/emulate.c:
```diff
@@ -90,6 +90,18 @@ static void rv_exception_default_handler(riscv_t *rv)
 RV_EXCEPTION_LIST
 #undef _
 
+#define RV_EXC_MISAGLIN_HANDLER(mask_or_pc, type, compress, IO)    \
```
Typo. Should be `RV_EXC_MISALIGN_HANDLER`.

jserv replied:
Add some comments for this macro.

jserv commented on src/emulate.c:
```diff
@@ -90,6 +90,19 @@ static void rv_exception_default_handler(riscv_t *rv)
 RV_EXCEPTION_LIST
 #undef _
 
+/* wrap load/store and insn misaligned handler */
```
You shall explain the parameters in details.

qwe661234 replied:
Updated

jserv reviewed (changes requested):
Integrate memory pool in conjunction of proposed memory operation changes.

jserv commented on src/emulate.c:
```diff
@@ -90,6 +90,25 @@ static void rv_exception_default_handler(riscv_t *rv)
 #undef _
 
+/* wrap load/store and insn misaligned handler
+ * @mask_or_pc: misaligned mask for load/store misaligned handler and pc for
```
Shorten the description for `mask_or_pc`.

jserv reviewed (changes requested):
Get rid of GPL'd TLSF. You MUST use the BSD license re-implementation instead. See #94 .
Original TLSF is licensed under GNU GPL, which raises more restrictions over MIT License. We can integrate BSD License components in rv32emu, but it is not allowed to relicense GPL'd ones into rv32emu. That is exactly what BSD re-implementation appears.

jserv commented:
Also, the original TLSF suffers from several security concerns. For this issue, we only need a simplified memory pool without integrating TLSF, where #94 emphasizes.

qwe661234 commented:
> Also, the original TLSF suffers from several security concerns. For this issue, we only need a simplified memory pool without integrating TLSF, where #94 emphasizes.

I believe that the memory pool with a fixed chunk size can fulfill our requirements, so I decided to integrate this type of memory pool into the file `io.c`.

jserv reviewed (changes requested):
Squash the commits along with comprehensive git commit messages. In addition, show the latest benchmark results.

jserv commented on src/emulate.c:
```diff
@@ -90,6 +90,24 @@ static void rv_exception_default_handler(riscv_t *rv)
+ * @mask_or_pc: mask for load/store and pc for insn misaligned handler.
+ * @type: type of misaligned handler
+ * @compress: compressed instruction or not
+ * @IO: misaligned handler is load/store or insn
```
Improve the writing for `@IO` entry.

jserv replied:
Ask ChatGPT to proofread.

jserv commented on src/io.c:
```diff
@@ -9,27 +9,25 @@
 static const uint32_t mask_lo = 0xffff;
 static const uint32_t mask_hi = ~(0xffff);
 
+struct mpool *mp;
```
Can you make it local? That is, declare this variable as `static`.

jserv commented on src/io.c:
```diff
@@ -125,23 +107,54 @@ void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
         if (!c) {
-            c = malloc(sizeof(chunk_t));
-            memset(c->data, 0, sizeof(c->data));
+            c = calloc(1, sizeof(chunk_t));
```
The region should be requested from memory pool.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,84 @@
+    struct mpool *next;
+} mpool_t;
+
+static void *get_mmap(long sz)
```
Check the file `src/elf.c` for `USE_MMAP`. At present, you can figure out Microsoft Windows counterpart or stop compilation (by placing `#error`).

Portability is an important consideration.

jserv commented on src/mpool.h:
```diff
@@ -0,0 +1,24 @@
+#include <sys/mman.h>
+#include <unistd.h>
+
+struct memchunk;
```
`struct memchunk` is implementation specific. Don't appear in this header file.

jserv commented on src/mpool.h:
```diff
@@ -0,0 +1,24 @@
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
```
Add proper `#pragma once`.

jserv commented on src/mpool.h:
```diff
@@ -0,0 +1,24 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
```
Minimize the header inclusion! Only keep the necessary ones.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,84 @@
+} memchunk_t;
+
+typedef struct mpool {
+    long chunk_ct;
```
`chunk_ct` is not a appropriate mnemonic.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,84 @@
+    long chunk_ct;
+    long page_count;
+    size_t chunk_size;
+    char *mempage;
```
Rename it to `mapped`, which stands for mapped memory region.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,84 @@
+typedef struct mpool {
+    long chunk_ct;
+    long page_count;
+    size_t chunk_size;
```
Check if `size_t` fits the range requirements for `chunk_ct` and `page_count`.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,84 @@
+
+mpool_t *mpool_create(size_t pool_size, size_t chunk_size)
+{
+    mpool_t *new_mp = malloc(sizeof(mpool_t));
```
Return `NULL` when `malloc` or `mmap` fails.

jserv commented on src/io.c:
```diff
@@ -125,23 +107,54 @@ void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
     }
 }
 
+void memory_write_w(memory_t *m, uint32_t addr, const uint8_t *src)
```
Use macro to generate the function bodies of `memory_write_{w,s,b}`.

jserv reviewed (changes requested):
Append `Close #122` at the end of git commit messages.

jserv commented on src/mpool.h:
```diff
@@ -0,0 +1,25 @@
+
+#include <stdlib.h>
+#include <string.h>
+#if !defined(_WIN32)
```
No, implementation specific statements should no appear in this header! Keep it simple and stupid.

jserv commented on src/mpool.h:
```diff
@@ -0,0 +1,20 @@
+ */
+#pragma once
+
+#include <stdlib.h>
```
Only `<stddef.h>` should be included here.

qwe661234 replied:
The gcc-12 compiler shows the warning when I only include `<stddef.h>`.

jserv replied:
I don't think so.
```
gcc-12 -c mpool.h -Wall
```
It compiles without warnings. The actual header inclusion should appear in C source files.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,109 @@
+    struct mpool *next;
+} mpool_t;
+
+static void *get_mmap(size_t sz)
```
Rename to `mem_arena` and add comments.

jserv commented on src/io.c:
```diff
@@ -9,27 +9,25 @@
-    memory_t *m = malloc(sizeof(memory_t));
-    memset(m->chunks, 0, sizeof(m->chunks));
+    memory_t *m = calloc(1, sizeof(memory_t));
+    mp = mpool_create(sizeof(chunk_t) << 5, sizeof(chunk_t));
```
You shall explain why shift 5 is here.

jserv commented on src/io.c:
```diff
@@ -85,63 +83,61 @@ uint32_t memory_ifetch(memory_t *m, uint32_t addr)
     }
 }
 
+#define memory_write(size, type)                                             \
```
Rename to `MEM_WRITE_IMPL`.

jserv reviewed (changes requested):
Rephrase the git commit message as following:
> After analyzing the data collected during the execution of the Dhrystone benchmark, it was discovered that the primary performance bottleneck lies in memory read/write operations. To address this issue, modifications were made to the implementation of memory read/write operations, eliminating unnecessary checks and replacing slow operations with more efficient ones. Additionally, a simple memory pool was integrated. It is now possible for the user to enable misaligned memory access by launching with option "--misalign".
> 
> The enhancements made to the memory I/O resulted in a significant improvement, as observed through performance results obtained by running Dhrystone.

jserv commented on src/mpool.h:
```diff
@@ -0,0 +1,20 @@
+
+struct mpool;
+
+struct mpool *mpool_create(size_t pool_size, size_t chunk_size);
```
Add informative comments for each function.

jserv commented on src/io.c:
```diff
@@ -9,27 +9,27 @@
-    memory_t *m = malloc(sizeof(memory_t));
-    memset(m->chunks, 0, sizeof(m->chunks));
+    memory_t *m = calloc(1, sizeof(memory_t));
+    /* sizeof(chunk_t) << 5 satisfies the memory requirement for all test cases
```
If `sizeof(chunk_t) << 5` does not fit, there should be some mechanism to extend.

jserv commented on src/mpool.c:
```diff
@@ -0,0 +1,109 @@
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
```
Add proper header inclusion here.

jserv commented on src/emulate.c:
```diff
@@ -90,6 +90,24 @@ static void rv_exception_default_handler(riscv_t *rv)
+     if (unlikely(insn_is_misaligned(rv->PC))))                    \
+    {                                                              \
+        rv->compressed = compress;                                 \
+        IIF(IO)                                                    \
```
Can you simplify as following?
```c
rv_except_##type##_misaligned(rv, IIF(IO)(addr, mask_or_pc));
```

qwe661234 committed:
Improve memory read/write

After analyzing the data collected during the execution of the Dhrystone
benchmark, it was discovered that the primary performance bottleneck lies
in memory read/write operations. To address this issue, modifications were
made to the implementation of memory read/write operations, eliminating
unnecessary checks and replacing slow operations with more efficient ones.
Additionally, a simple memory pool was integrated. It is now possible for
the user to enable misaligned memory access by launching with option "--misalign".

The enhancements made to the memory I/O resulted in a significant improvement,
as observed through performance results obtained by running Dhrystone.

|    Test    |    66200d0   |   improvement   |Speedup|
|------------+--------------+-----------------+-------|
| Dhrystone  | 815 DMIPS    |   1245 DMIPS    | +52.7%|

Close #122

jserv reviewed (changes requested):
Append a test item which triggers option `--misalign` in GitHub Actions. That is, we shall regularly validate the case when misaligned memory access is allowed for CI pipeline.

qwe661234 committed:
Add misaligned  memory access test item
